src/common/providers/database.ts

```diff
@@ -0,0 +1,344 @@
+// SOFTWARE.
+
+import * as firebase from 'firebase-admin';
+import { joinPath, /*normalizePath,*/ pathParts } from '../../utilities/path';
+
+/**
+ * Interface representing a Firebase Realtime Database data snapshot.
+ */
+export class DataSnapshot {
+  public instance: string;
+
+  /** @hidden */
+  private _ref: firebase.database.Reference;
+
+  /** @hidden */
+  private _path: string;
+
+  /** @hidden */
+  private _data: any;
+
+  /** @hidden */
+  private _childPath: string;
+
+  constructor(
+    data: any,
+    path?: string, // path will be undefined for the database root
+    private app?: firebase.app.App,
+    instance?: string
+  ) {
+    if (app?.options?.databaseURL?.startsWith('http:')) {
+      // In this case we're dealing with an emulator
+      this.instance = app.options.databaseURL;
+    } else if (instance) {
+      // SDK always supplies instance, but user's unit tests may not
+      this.instance = instance;
+    } else if (app) {
+      this.instance = app.options.databaseURL;
+    } else if (process.env.GCLOUD_PROJECT) {
+      this.instance =
+        'https://' +
+        process.env.GCLOUD_PROJECT +
+        '-default-rtdb.firebaseio.com';
+    }
+
+    this._path = path;
+    this._data = data;
+  }
+
+  /**
+   * Returns a [`Reference`](/docs/reference/admin/node/admin.database.Reference)
+   * to the Database location where the triggering write occurred. Has
+   * full read and write access.
+   */
+  get ref(): firebase.database.Reference {
+    if (!this.app) {
+      // may be unpopulated in user's unit tests
+      throw new Error(
+        'Please supply a Firebase app in the constructor for DataSnapshot' +
+          ' in order to use the .ref method.'
+      );
+    }
+    if (!this._ref) {
+      this._ref = this.app.database(this.instance).ref(this._fullPath());
+    }
+    return this._ref;
+  }
+
+  /**
+   * The key (last part of the path) of the location of this `DataSnapshot`.
+   *
+   * The last token in a Database location is considered its key. For example,
+   * "ada" is the key for the `/users/ada/` node. Accessing the key on any
+   * `DataSnapshot` will return the key for the location that generated it.
+   * However, accessing the key on the root URL of a Database will return `null`.
+   */
+  get key(): string {
+    // const last = _.last(pathParts(this._fullPath()));
+    const segments = pathParts(this._fullPath());
+    const last = segments[segments.length - 1];
+    return !last || last === '' ? null : last;
+  }
+
+  /**
+   * Extracts a JavaScript value from a `DataSnapshot`.
+   *
+   * Depending on the data in a `DataSnapshot`, the `val()` method may return a
+   * scalar type (string, number, or boolean), an array, or an object. It may also
+   * return `null`, indicating that the `DataSnapshot` is empty (contains no
+   * data).
+   *
+   * @return The DataSnapshot's contents as a JavaScript value (Object,
```

Comment:
This should be a literal, or maybe just "The snapshot's contents?"   A snapshot can be a non-capitalized, generic noun IMO.

Reply:
Fixed